commit 4db8b614926ebee81a785cb070669c437cad3201
Author: Ryan Zezeski <rpz@joyent.com>
Date:   2019-07-25T12:06:28-06:00 (2 months ago)
    
    OS-7904 simnet has bogus mi_tx_cksum_flags
    OS-7905 mac_tx() is too eager to emulate hardware offloads

diff --git a/usr/src/uts/common/io/mac/mac_client.c b/usr/src/uts/common/io/mac/mac_client.c
index d2444094a9..5c60bc9563 100644
--- a/usr/src/uts/common/io/mac/mac_client.c
+++ b/usr/src/uts/common/io/mac/mac_client.c
@@ -3673,7 +3673,7 @@ mac_tx(mac_client_handle_t mch, mblk_t *mp_chain, uintptr_t hint,
 
 			mp->b_next = NULL;
 
-			if (needed != 0) {
+			if ((needed & (HCK_TX_FLAGS | HW_LSO_FLAGS)) != 0) {
 				mac_emul_t emul = 0;
 
 				if (needed & HCK_IPV4_HDRCKSUM)
diff --git a/usr/src/uts/common/io/mac/mac_provider.c b/usr/src/uts/common/io/mac/mac_provider.c
index 99f2f3431e..2ad4485a3c 100644
--- a/usr/src/uts/common/io/mac/mac_provider.c
+++ b/usr/src/uts/common/io/mac/mac_provider.c
@@ -21,7 +21,7 @@
 
 /*
  * Copyright (c) 2008, 2010, Oracle and/or its affiliates. All rights reserved.
- * Copyright 2018 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  * Copyright 2017 OmniTI Computer Consulting, Inc. All rights reserved.
  */
 
@@ -1008,7 +1008,18 @@ mac_pdata_update(mac_handle_t mh, void *mac_pdata, size_t dsize)
 void
 mac_capab_update(mac_handle_t mh)
 {
-	/* Send MAC_NOTE_CAPAB_CHG notification */
+	mac_impl_t *mip = (mac_impl_t *)mh;
+
+	/*
+	 * Depending on which capabs changed, the Tx checksum flags
+	 * may also need updating.
+	 */
+	mip->mi_tx_cksum_flags = mac_features_to_flags(mh);
+
+	/*
+	 * Send a MAC_NOTE_CAPAB_CHG notification to alert upstream
+	 * clients to renegotiate capabilities.
+	 */
 	i_mac_notify((mac_impl_t *)mh, MAC_NOTE_CAPAB_CHG);
 }
 
diff --git a/usr/src/uts/common/io/simnet/simnet.c b/usr/src/uts/common/io/simnet/simnet.c
index b169e519b1..3653ed12f0 100644
--- a/usr/src/uts/common/io/simnet/simnet.c
+++ b/usr/src/uts/common/io/simnet/simnet.c
@@ -1385,6 +1385,12 @@ simnet_m_setprop(void *arg, const char *name, mac_prop_id_t num,
 		break;
 	}
 
+	/*
+	 * We may have modified the configuration of hardware
+	 * offloads. Make sure to renegotiate capabilities with the
+	 * upstream clients.
+	 */
+	mac_capab_update(sdev->sd_mh);
 	return (err);
 }
 
diff --git a/usr/src/uts/common/sys/pattr.h b/usr/src/uts/common/sys/pattr.h
index 587a51f0aa..d4df270967 100644
--- a/usr/src/uts/common/sys/pattr.h
+++ b/usr/src/uts/common/sys/pattr.h
@@ -21,7 +21,7 @@
 /*
  * Copyright 2010 Sun Microsystems, Inc.  All rights reserved.
  * Use is subject to license terms.
- * Copyright 2018 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 #ifndef _SYS_PATTR_H
@@ -98,6 +98,8 @@ typedef struct pattr_hcksum_s {
 
 #define	HCK_FLAGS		(HCK_IPV4_HDRCKSUM | HCK_PARTIALCKSUM |	\
 				HCK_FULLCKSUM | HCK_FULLCKSUM_OK)
+#define HCK_TX_FLAGS		(HCK_IPV4_HDRCKSUM | HCK_PARTIALCKSUM | \
+				HCK_FULLCKSUM)
 /*
  * Extended hardware offloading flags that also use hcksum_flags
  */
